name: common-deploy

on:
  workflow_call:

jobs:
  deploy:
    name: デプロイ
    runs-on: ubuntu-latest
    timeout-minutes: 15

    steps:
      - name: Checkout
        uses: actions/checkout@v4

      - name: Setup Node.js 20
        uses: actions/setup-node@v4
        with:
          node-version: 20

      - name: Install Clasp
        run: |
          npm init -y
          npm install clasp -g

      - name: Create .clasprc.json
        env:
          CLASPRC_JSON: ${{ secrets.CLASPRC_JSON }}
        run: |
          echo $CLASPRC_JSON > ~/.clasprc.json

      - name: Output summary
        run: |
          CLASP_JSON=`cat ./.clasp.json`
          MANIFEST=`cat ./src/appsscript.json`
          echo "## .clasprc.json" >> $GITHUB_STEP_SUMMARY
          if [ -e ~/.clasprc.json ]; then
            echo "存在します"
          else
            echo "存在しません"
          fi
          echo "\`\`\`" >> $GITHUB_STEP_SUMMARY
          echo -e "$CLASPRC_JSON" >> $GITHUB_STEP_SUMMARY
          echo "\`\`\`" >> $GITHUB_STEP_SUMMARY
          echo "## .clasp.json" >> $GITHUB_STEP_SUMMARY
          echo "\`\`\`" >> $GITHUB_STEP_SUMMARY
          echo -e "$CLASP_JSON" >> $GITHUB_STEP_SUMMARY
          echo "\`\`\`" >> $GITHUB_STEP_SUMMARY
          echo "## appsscript.json" >> $GITHUB_STEP_SUMMARY
          echo "\`\`\`" >> $GITHUB_STEP_SUMMARY
          echo -e "$MANIFEST" >> $GITHUB_STEP_SUMMARY
          echo "\`\`\`" >> $GITHUB_STEP_SUMMARY

      # clasp push実行
      # 1.clasp pushを実行し、その標準出力＋標準エラー出力を変数にセットする
      # 2.変数を出力する
      # 3.$GITHUB_OUTPUTにセット（複数行用の書式）
      - name: Clasp push
        id: clasp_push
        run: |
          RESULT=`clasp push -f 2>&1`
          echo -e "$RESULT"
          {
            echo "result<<EOF"
            echo "$RESULT"
            echo "EOF"
          }  >> $GITHUB_OUTPUT

      # clasp pushが失敗した場合にjobをエラー終了させるためのstep
      - name: Failed clasp push
        # clasp pushの結果に「Push failed」が含まれている、もしくは、「Error retrieving」で始まる場合は失敗したと判断
        if: ${{ contains(steps.clasp_push.outputs.result, 'Push failed') || startsWith(steps.clasp_push.outputs.result, 'Error retrieving') }}
        run: |
          exit 1
